src/validators/union.rs

```diff
@@ -134,16 +136,20 @@ impl UnionValidator {
                     _ => {
                         // success should always have an exactness
                         debug_assert_ne!(state.exactness, None);
+
                         let new_exactness = state.exactness.unwrap_or(Exactness::Lax);
-                        // if the new result has higher exactness than the current success, replace it
-                        if success
-                            .as_ref()
-                            .map_or(true, |(_, current_exactness)| *current_exactness < new_exactness)
-                        {
-                            // TODO: is there a possible optimization here, where once there has
-                            // been one success, we turn on strict mode, to avoid unnecessary
-                            // coercions for further validation?
-                            success = Some((new_success, new_exactness));
+                        let new_fields_set = state.fields_set_count;
```

Comment:
I think we probably need to reset `fields_set_count` between each branch?

Propagating on the state like this potentially has surprising effects like `list[ModelA] | list[ModelB]` probably now gets chosen based on whether the last entry fitted `ModelA` or `ModelB` better. We could deal with that by explicitly clearing `num_fields_set` in compound validators like list validators.

Overall I'm fine with this approach of using state, I think either this or `num_fields` (and `exactness`) all are fiddly to implement and will likely just need lots of hardening via test cases.

Reply:
Done!

Reply:
Behavior looks good for the list case you've mentioned above:

```py
from pydantic import BaseModel, TypeAdapter

class ModelA(BaseModel):
    a: int

class ModelB(ModelA):
    b: int

ta = TypeAdapter(list[ModelA] | list[ModelB])

print(repr(ta.validate_python([{'a': 1}, {'a': 2}])))
#> ModelA(a=1), ModelA(a=2)]

print(repr(ta.validate_python([{'a': 1, 'b': 2}, {'a': 2, 'b': 3}])))
#> [ModelB(a=1, b=2), ModelB(a=2, b=3)]

print(repr(ta.validate_python([{'a': 1, 'b': 2}, {'a': 2}])))
#> [ModelA(a=1), ModelA(a=2)]

print(repr(ta.validate_python([{'a': 1}, {'a': 2, 'b': 3}])))
#> [ModelA(a=1), ModelA(a=2)]
```

I think this is because I'm now clearing the relevant parts of the state, and `fields_set_count` is by default `None`, in which case we only consider `exactness`.